### ISSUE TITLE
mockito does not install with flutter_driver on stable

### ISSUE THREAD

**Issue by IchordeDionysos** [flutter/flutter#77282]
## Steps to Reproduce

1. Update to Flutter 2.
2. Add `mockito: 5.0.0` (w/ null-safety).
3. Add `flutter_driver: sdk: flutter`
3. Run `flutter pub get`

**Expected results:** `flutter pub get` runs successfully.

**Actual results:** `fluter pub get` complains that these two packages are not compatible due to `crypto`.

<details>
  <summary>Logs</summary>

```
Because every version of flutter_driver from sdk depends on crypto 2.1.5 and analyzer >=1.0.0 depends on crypto ^3.0.0, flutter_driver from sdk is incompatible with analyzer >=1.0.0.
And because mockito >=5.0.0 depends on analyzer ^1.0.0, flutter_driver from sdk is incompatible with mockito >=5.0.0.
So, because firebase_analytics depends on both mockito ^5.0.0 and flutter_driver any from sdk, version solving failed.
Running "flutter pub get" in firebase_analytics...                      
pub get failed (1; So, because firebase_analytics depends on both mockito ^5.0.0 and flutter_driver any from sdk, version solving failed.)
```

```
[✓] Flutter (Channel stable, 2.0.0, on macOS 11.2.2 20D80 darwin-x64, locale de-DE)
    • Flutter version 2.0.0 at /...
    • Framework revision 60bd88df91 (27 hours ago), 2021-03-03 09:13:17 -0800
    • Engine revision 40441def69
    • Dart version 2.12.0

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/_/Library/Android/sdk
    • Platform android-30, build-tools 29.0.2
    • Java binary at: /Users/_/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/201.7042882/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Users/_/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/201.7042882/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[✓] VS Code (version 1.52.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.18.1

[✓] Connected device (2 available)

• No issues found!
```

</details>

After a bit of investigation, I've found that the @dnfield did already [bump `crypto` to `v3.0.0`](https://github.com/flutter/flutter/commit/d1538320d21c67ac5f90ae1b1b4ca6c76307aebe#diff-745da1477f35a8b50e9abf8ad376bfae7a344aafac5a48628a520fe4204c7715R29) over a month ago, but that change hasn't yet landed in [beta](https://github.com/flutter/flutter/blob/beta/packages/flutter_driver/pubspec.yaml#L31) nor [stable](https://github.com/flutter/flutter/blob/stable/packages/flutter_driver/pubspec.yaml#L31).

This kinda blocks the migration to null-safety for basically 3 months (until the next stable release) 🤔
So I am wondering if this is something that can/will land earlier than the normal release schedule (via a cherry-pick) or not 🤔

**Comment by IchordeDionysos** [comment]
By the way, I've run into this while trying to migrate `firebase_analytics` to null safety :)

**Comment by dnfield** [comment]
This should be fixed on dev/master, but would require a cherry pick to fix on stable. @christopherfujino could probably help determine that. I'm out on some baby leave for now and don't have much bandwidth.

**Comment by TahaTesser** [comment]
Hi @IchordeDionysos 
Thanks for fling the issue, I can confirm this is working fine on the `dev` and `master`.

`stable` and `beta`
```bash
Because every version of flutter_driver from sdk depends on crypto 2.1.5 and analyzer >=1.0.0 depends on crypto ^3.0.0, flutter_driver from sdk is incompatible with analyzer >=1.0.0.
And because mockito 5.0.0 depends on analyzer ^1.0.0 and no versions of mockito match >5.0.0 <6.0.0, flutter_driver from sdk is incompatible with mockito ^5.0.0.
So, because triage depends on both mockito ^5.0.0 and flutter_driver any from sdk, version solving failed.
```

`dev` and `master`
```
Running "flutter pub get" in stable_flutter...                   1,378ms
```


<details> 
<summary>flutter doctor -v</summary> 

```bash
[✓] Flutter (Channel stable, 2.0.1, on macOS 11.2.2 20D80 darwin-x64, locale
    en-GB)
    • Flutter version 2.0.1 at /Users/tahatesser/Code/flutter_stable
    • Framework revision c5a4b4029c (15 hours ago), 2021-03-04 09:47:48 -0800
    • Engine revision 40441def69
    • Dart version 2.12.0

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Volumes/Extreme/SDK
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = /Volumes/Extreme/SDK
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Volumes/Extreme/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)

[✓] VS Code (version 1.54.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.20.0

[✓] Connected device (3 available)
    • SM M025F (mobile) • R9ZR205XX0A • android-arm    • Android 10 (API 29)
    • macOS (desktop)   • macos       • darwin-x64     • macOS 11.2.2 20D80
      darwin-x64
    • Chrome (web)      • chrome      • web-javascript • Google Chrome
      88.0.4324.192

• No issues found!
```

```bash
[✓] Flutter (Channel beta, 2.0.1, on macOS 11.2.2 20D80 darwin-x64, locale
    en-GB)
    • Flutter version 2.0.1 at /Users/tahatesser/Code/flutter_beta
    • Framework revision c5a4b4029c (15 hours ago), 2021-03-04 09:47:48 -0800
    • Engine revision 40441def69
    • Dart version 2.12.0

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Volumes/Extreme/SDK
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = /Volumes/Extreme/SDK
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Volumes/Extreme/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)

[✓] VS Code (version 1.54.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.20.0

[✓] Connected device (3 available)
    • SM M025F (mobile) • R9ZR205XX0A • android-arm    • Android 10 (API 29)
    • macOS (desktop)   • macos       • darwin-x64     • macOS 11.2.2 20D80
      darwin-x64
    • Chrome (web)      • chrome      • web-javascript • Google Chrome
      88.0.4324.192

• No issues found!
```

</details>

Thank you

**Comment by samandmoore** [comment]
Would be great to get this cherry picked to stable if it's not a huge lift. 

This is presumably preventing a decent number of folks from upgrading to flutter 2; I know it's the last item blocking my team.

**Comment by christopherfujino** [comment]
> Would be great to get this cherry picked to stable if it's not a huge lift.
> 
> This is presumably preventing a decent number of folks from upgrading to flutter 2; I know it's the last item blocking my team.

Can you clarify how this is blocking your team from migrating to Flutter 2? Are you using a pre 2.0 version of Flutter with a null-safe mockito?

**Comment by samandmoore** [comment]
> Can you clarify how this is blocking your team from migrating to Flutter 2? Are you using a pre 2.0 version of Flutter with a null-safe mockito?

sorry to be vague!

so, my issue isn't actually not mockito specific, but it's related flutter_driver and incompatible shared dependency constraints (so the same underlying type of issue as reported here). 

the specific issue is here:

```
Because lottie 1.0.1 depends on archive ^3.0.0 and no versions of lottie match >1.0.1 <2.0.0, lottie ^1.0.1 requires archive ^3.0.0.

And because every version of flutter_driver from sdk depends on archive 2.0.13 and every version of betterment_flutter from path depends on lottie ^1.0.1, flutter_driver from sdk is incompatible with betterment_flutter from path.

So, because app_standalone depends on both betterment_flutter from path and flutter_driver any from sdk, version solving failed.
pub get failed (1; So, because app_standalone depends on both betterment_flutter from path and flutter_driver any from sdk, version solving failed.)
```

**Comment by samandmoore** [comment]
I'm not sure if it's safe to work-around this with an entry in `dependency_overrides`, but by specifying overrides for archive and crypto, I can get `pub get` to work.

**Comment by IchordeDionysos** [comment]
@christopherfujino the underlying issue is that `flutter_driver` is [not yet null-safe in beta/stable](https://api.flutter.dev/flutter/flutter_driver/flutter_driver-library.html) (no null-safety badge).

And as these libraries are shipped with Flutter, this basically means that the migration to full null-safety is blocked until the next stable release of Flutter.

To put this into perspective:
This technically prevents `firebase_analytics`, and others from migrating to null-safety as they rely on `flutter_driver` ...

**Comment by marcglasberg** [comment]
I have opened a specific issue to track moving null-safe flutter_driver to the stable channel: https://github.com/flutter/flutter/issues/80293

**Comment by jonahwilliams** [comment]
We're not going to consider cherry picks to stable and beta for null safe driver or integration test. We evaluated this when creating the release branch for flutter 2 and concluded that it was too risky. Null safe driver and integration test will be available in the next beta/stable release

**Comment by marcglasberg** [comment]
So it's going to take months? Maybe you should have waited to release Flutter 2 if it was not ready. Lots of pressure for package developers to implement null-safety, but Flutter itself didn't do its homework. I have packages people asked me to downgrade dependencies because of flutter_driver.

**Comment by dnfield** [comment]
I agree this is not an ideal situation, and I really wish we had been able to get this one in time for the 2.0 branch cut. That said consider the following:

- The current beta channel of flutter offers a NNBD Flutter Driver
- The next release version will be out relatively soon, and will contain this change. Per our release process, we update roughly once a quarter, and we'll be due for one before the end of this quarter :) (https://medium.com/flutter/flutter-spring-2020-update-f723d898d7af)
- Clients that are facing issues with particular packages can try using `dependency_overrides` in the mean time. for example:

```yaml

dependency_overrides:
  crypto: ^3.0.0
```

Many of the NNBD packages were released with breaking semver changes "just in case," but are compatible with the previous major version - I know this is true for crypto in particular.

Ideally we would have been able to migrate all of flutter_driver and its upstream dependencies before the 2.0 branch cut. Unfortunately, that work had a long tail of interrelated dependencies and it was determined that it wasn't worth delaying the branch cut over that work. Enough of it had been migrated that NNBD tests could be written, and work arounds are possible for the relatively small number of customers who would be impacted by things like needing dependency overrides etc.

**Comment by marcglasberg** [comment]
Ok, thanks for the explanation. Flutter's eager to sacrifice a bit of stability/feature completeness, to keep release dates. It's just that I would have preferred the other way around. I would have delayed the branch cut. Thanks again!

**Comment by Valentin-Seehausen** [comment]
Hey, thanks for the explanation on this thread. For a new project with null safety this effectively means that integration tests are not really feasible, right?

**Comment by dnfield** [comment]
Yes, or use the beta or dev channels for now.

**Comment by FrantisekGazo** [comment]
any timeline on fixing this in the stable channel?

**Comment by IchordeDionysos** [comment]
Flutter 2.0 hit stable on March 3rd.
Stable channel is released about once a quarter.
So it should land around the June 3rd, earlier or later (very likely)
There might be a delay that I can't comment on.

**Comment by IchordeDionysos** [comment]
This is now fixed with the Flutter 2.2 release 😍
